gio/src/rust_async_result.rs

```diff
@@ -0,0 +1,163 @@
+    /// [`crate::Task`] in situations where the future does not need to be spawned on a new thread.
+    pub struct RustAsyncResult(ObjectSubclass<RustAsyncResultPrivate>)
+        @implements AsyncResult;
+}
+
+impl RustAsyncResult {
+    // rustdoc-stripper-ignore-next
+    /// Spawns `fut` on the thread default [`MainContext`](glib::MainContext) and calls `callback`
+    /// when it completes. In order, the arguments passed to the callback will be:
+    ///  - `source_object`
+    ///  - a new `RustAsyncResult` holding the value returned from `fut`
+    ///  - `user_data`
+    ///
+    /// If `cancellable` is not `None`, then the callback will be called either when the future
+    /// completes or when `cancellable` is cancelled. If cancelled, the result will hold a
+    /// [`glib::Error`] with the code [`crate::IOErrorEnum::Cancelled`].
+    ///
+    /// `source_tag` is for debugging purposes and should be a unique pointer tagging the result.
+    /// Usually, it is given as the function pointer of the calling C method.
+    ///
+    /// # Safety
+    ///
+    /// The convention in the C API is for `callback` to have its closure data stored in
+    /// `user_data`, and then dereference the pointer when `callback` is called. Callers should
+    /// pass `callback` and `user_data` directly from C and not manipulate them in any way.
+    pub unsafe fn for_spawn<O, C, F, R>(
+        source_object: &O,
+        source_tag: glib::Pointer,
+        cancellable: Option<&C>,
+        callback: ffi::GAsyncReadyCallback,
+        user_data: glib::Pointer,
+        fut: F,
+    ) where
+        O: IsA<Object>,
+        C: IsA<Cancellable>,
+        F: Future<Output = Result<R, glib::Error>> + 'static,
+        R: ToValue,
+    {
+        let callback = callback.map(|callback| {
+            let obj = source_object.upcast_ref::<Object>().clone();
+            move |value| {
+                let result = Object::new::<Self>(&[]).unwrap();
+                result
+                    .imp()
+                    .inner
+                    .set(Inner {
+                        source_object: obj.clone(),
+                        user_data,
+                        source_tag,
+                    })
+                    .unwrap();
+                result.imp().result.replace(Some(value));
+                callback(
+                    obj.to_glib_none().0,
+                    result.upcast_ref::<AsyncResult>().to_glib_none().0,
+                    user_data,
+                );
+            }
+        });
+        let cancellable = cancellable.map(|c| c.upcast_ref::<Cancellable>()).cloned();
+        glib::MainContext::ref_thread_default().spawn_local(async move {
+            if let Some(callback) = callback {
+                let res = if let Some(cancellable) = &cancellable {
+                    futures_util::select! {
+                        res = fut.fuse() => res,
+                        _ = cancellable.future().fuse() => Err(
```

Comment:
You should move both fuse() calls outside the `select!` here and also below. The futures have to be completely constructed outside the macro.

But also as you only have two futures here maybe just use https://docs.rs/futures/latest/futures/future/fn.select.html ?